### DISCUSSION TITLE
What is the best way to tile an image with another image?

### DISCUSSION THREAD

**Discussion by brandoncc** [libvips/libvips#2428] [Q&A] [answered]
If I have image1 that is 100x100 and image2 which is 750x750, what is the most efficient way to tile image2 with 7 and a half of image1 in each direction? joinarray says it accepts a grid, but it doesn't seem to accept more than one row in the grid?

This is the version I currently wrote, which has a big O that leaves something to be desired, and therefore is slow:

```ruby
require 'down'
require 'vips'

width = 750
height = 750

image1_file = Down.download 'https://placeimg.com/100/100/cat.jpg'
image1 = Vips::Image.new_from_file(image1_file.path)
image1_row = image1

while image1_row.width < width
  image1_row = image1_row.join image1, 'horizontal'
end

image2 = image1_row

while image2.height < height
  image2 = image2.join image1_row, 'vertical'
end

image2 = image2.crop 0, 0, height, width

```

Here is the result of that code:

![image](https://user-images.githubusercontent.com/543973/132741755-e40893fc-e16d-470a-a02b-d9f6cd1181c9.png)

Thanks!

**Comment by jcupitt** [comment, marked as answer]
Hi @brandoncc,

You need `replicate`:

https://libvips.github.io/libvips/API/current/libvips-conversion.html#vips-replicate

Something like:

```ruby
target_width = 100000
target_height = 100000
image = Vips::Image.new_from_file filename
huge = image.replicate image.width / target_width + 1, image.height / target_height + 1
huge = huge.crop 0, 0, target_width, target_height
```

**Comment by jcupitt** [comment]
You're right, you could use `arrayjoin`, but `replicate` should be a bit faster.

**Reply by jcupitt** [reply]
`arrayjoin` version:

```ruby
#!/usr/bin/ruby

require 'vips'

image = Vips::Image.new_from_file ARGV[0]
huge = Vips::Image.arrayjoin [image] * 25, across: 5
huge.thumbnail_image(512).write_to_file(ARGV[1])
```

**Reply by brandoncc** [reply]
Ohhhh, so it is a list of images and you tell it the width of the row. That wasn't clear in the docs, but that could totally be my fault.

**Reply by jcupitt** [reply]
The downside of automatically generated docs :(

https://www.rubydoc.info/gems/ruby-vips/Vips%2FImage.arrayjoin

**Reply by brandoncc** [reply]
I've never thought to look at that site for ruby-vips docs, I always go to the libvips docs site and then have to figure out exactly the right way to convert to ruby usage (which I'm getting pretty good at). Thanks for dropping even more knowledge on me :-)

**Comment by brandoncc** [comment]
Thank you! Sorry for bothering you, I missed `replicate` when I scrolled down that page in the docs.

**Reply by jcupitt** [reply]
Oh. no problem. Replicate is useful for any kind of image tiling, eg.:

```ruby
require 'vips'

image = Vips::Image.new_from_file ARGV[0]
hor_pair = image.join image.fliphor, :horizontal
tile = hor_pair.join hor_pair.flipver, :vertical
huge = tile.replicate 3, 3
huge.thumbnail_image(512).write_to_file(ARGV[1])
```

To make a seamless mosaic, eg.:

![x](https://user-images.githubusercontent.com/580843/132750117-e80f341a-4fd1-415e-8c27-905940486d08.jpg)

There should probably be `joinhor` and `joinver` convenience methods.

**Reply by brandoncc** [reply]
This is really interesting, and might be very useful to me. Thanks!
